osu.Game/Scoring/ScoreInfo.cs

```diff
@@ -3,191 +3,235 @@
+        public List<HitEvent> HitEvents { get; set; } = new List<HitEvent>();
+
+        public ScoreInfo DeepClone()
+        {
+            var clone = (ScoreInfo)MemberwiseClone();
 
-        [NotMapped]
+            clone.Statistics = new Dictionary<HitResult, int>(clone.Statistics);
+
+            return clone;
+        }
+
+        [Ignored]
         public bool Passed { get; set; } = true;
 
-        public RulesetInfo Ruleset { get; set; }
+        public int Combo { get; set; }
 
-        private APIMod[] localAPIMods;
+        /// <summary>
+        /// The position of this score, starting at 1.
+        /// </summary>
+        [Ignored]
+        public int? Position { get; set; } // TODO: remove after all calls to `CreateScoreInfo` are gone.
 
-        private Mod[] mods;
+        [Ignored]
+        public LocalisableString DisplayAccuracy => Accuracy.FormatAccuracy();
 
-        [NotMapped]
+        /// <summary>
+        /// Whether this <see cref="EFScoreInfo"/> represents a legacy (osu!stable) score.
+        /// </summary>
+        [Ignored]
+        public bool IsLegacyScore => Mods.OfType<ModClassic>().Any();
+
+        // Used for database serialisation/deserialisation.
+        [MapTo("Mods")]
+        public string ModsJson { get; set; } = string.Empty;
+
+        [Ignored]
         public Mod[] Mods
         {
             get
             {
-                var rulesetInstance = Ruleset?.CreateInstance();
-                if (rulesetInstance == null)
-                    return mods ?? Array.Empty<Mod>();
-
                 Mod[] scoreMods = Array.Empty<Mod>();
 
                 if (mods != null)
                     scoreMods = mods;
                 else if (localAPIMods != null)
-                    scoreMods = APIMods.Select(m => m.ToMod(rulesetInstance)).ToArray();
+                    scoreMods = APIMods.Select(m => m.ToMod(Ruleset.CreateInstance())).ToArray();
 
                 return scoreMods;
             }
             set
             {
                 localAPIMods = null;
                 mods = value;
+                ModsJson = JsonConvert.SerializeObject(APIMods);
             }
         }
 
         // Used for API serialisation/deserialisation.
-        [NotMapped]
+        [Ignored]
         public APIMod[] APIMods
         {
             get
             {
-                if (localAPIMods != null)
-                    return localAPIMods;
+                if (localAPIMods == null)
+                {
+                    // prioritise reading from realm backing
+                    if (!string.IsNullOrEmpty(ModsJson))
+                        localAPIMods = JsonConvert.DeserializeObject<APIMod[]>(ModsJson);
 
-                if (mods == null)
-                    return Array.Empty<APIMod>();
+                    // then check mods set via Mods property.
+                    if (mods != null)
+                        localAPIMods = mods.Select(m => new APIMod(m)).ToArray();
+                }
```

Comment:
You've mentioned `Statistics` being quite ugly in the OP but this also needs to go too eventually. It's the same sort of badness.

Reply:
For sure.

I've done a basic clean-up pass on these to bring a bit more sanity to the variable naming (and make them both match as close as possible to each other in structure).